### PR TITLE
drivers: sensor: stm32_temp: Fix + add support for STM32G0B1

### LINKED ISSUE #41742: stm32g0: stm32_temp: not working
I'm trying the [stm32_temp](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/adc_stm32.c) driver on my G0B1RE custom board.

image

This is the characteristics from G0B1 datasheet:
image

My devicetree looks like this
```
/ {
	model = "GTSB IVM21";
	compatible = "gtsb,ivm21";

	chosen {
		zephyr,console = &usart2;
		zephyr,shell-uart = &usart2;
		zephyr,uart-mcumgr = &usart2;
		zephyr,sram = &sram0;
		zephyr,flash = &flash0;
		zephyr,code-partition = &slot0_partition;
	};
...
	vmain {
		compatible = "voltage-divider";
		io-channels = <&adc1 0>;
		output-ohms = <51000>;
		full-ohms = <(51000 + 51000)>;
	};

	stm-temp {
		label = "DIE_TEMP";
		compatible = "st,stm32-temp";
		io-channels = <&adc1 12>;
		avgslope = <25>;
		v25 = <760>;
		status = "okay";
	};
};
...
&adc1 {
	pinctrl-0 = <&adc1_in0_pa0 &adc1_in1_pa1 &adc1_in2_pa2 &adc1_in3_pa3>;
	status = "okay";
};
```

And I added these into my defconfig
```
CONFIG_SENSOR=y
CONFIG_STM32_TEMP=y
```

with no other changes to my existing application, apart from the devicetree & Kconfig changes, I found that my application fails to read the battery voltage, the channel setup fails here everytime:

image

I then changed the stm32_temp driver [init](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/stm32_temp/stm32_temp.c#L104) to: (I'm not sure what am I doing, just trial and error)

```c
	*accp = (struct adc_channel_cfg){
		.gain = ADC_GAIN_1,
		.reference = ADC_REF_INTERNAL,
		.acquisition_time = ADC_ACQ_TIME_DEFAULT,
		.channel_id = data->channel,
		.differential = 0
	};
```

Then, my application is able to read the ADC properly.

After that, I copied the stm32_temp sample code to my main() and found that the code stuck in the [calibration step](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/adc_stm32.c#L342), I then edited the stm32_temp driver so that the [calibration is set to false](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/stm32_temp/stm32_temp.c#L116).

```c
	*asp = (struct adc_sequence){
		.channels = BIT(data->channel),
		.buffer = &data->sample_buffer,
		.buffer_size = sizeof(data->sample_buffer),
		.resolution = 12,
		.calibrate = false,
	};
```

And now I'm able to read something, but it clearly doesn't work:
```
Current temperature: -172.-199996 °C
Current temperature: -172.-199996 °C
Current temperature: -171.-800003 °C
Current temperature: -171.-800003 °C
```

I'm not sure if it has negative temperature coefficient, this is what I get if I enabled `ntc`:
```
Current temperature: 221.399993 °C
Current temperature: 222.199996 °C
Current temperature: 221.399993 °C
Current temperature: 222.199996 °C
```

I do know that the formula in stm32_temp isn't suppose to "just works", as it uses v25, and I think I need v30 (so probably need to change this [line](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/stm32_temp/stm32_temp.c#L76) to += 30?)

I'm on Zephyr 2.7.0 RC3, `CONFIG_PM` is enabled.

[15. ADC.pdf](https://github.com/zephyrproject-rtos/zephyr/files/7854188/15.ADC.pdf)
[stm32g0b1re.pdf](https://github.com/zephyrproject-rtos/zephyr/files/7854195/stm32g0b1re.pdf)

### PR DESCRIPTION
**Pull request by ycsin**
Remove unnecessary calibration, fix for `stm32_temp`, add support for STM32G0B1

STM32G0B1 datasheet:
image

I think the configurations should be similar across G0 family, but the TS_CAL2 seems to be missing from other series (at least G0B0 & G070), so I decided to leave them out from this PR for now.

G0B0:
image

Log output from g0b1 custom board, integrated the [stm32_temp sample](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/sensor/stm32_temp_sensor/src/main.c) into my custom application that samples battery voltages:

```
[00:39:32.226,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2563|195] [4128 mV| 314 mV]
[00:39:42.327,000] <inf> component_power: Current temperature: 31.896545 °C
[00:39:42.327,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2563|192] [4128 mV| 308 mV]
[00:39:52.428,000] <inf> component_power: Current temperature: 31.580459 °C
[00:39:52.429,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2563|191] [4128 mV| 306 mV]
[00:40:02.530,000] <inf> component_power: Current temperature: 31.580459 °C
[00:40:02.530,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2562|193] [4128 mV| 310 mV]
[00:40:12.632,000] <inf> component_power: Current temperature: 31.896545 °C
[00:40:12.632,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2562|193] [4128 mV| 310 mV]
[00:40:22.733,000] <inf> component_power: Current temperature: 32.212631 °C
[00:40:22.733,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2561|193] [4126 mV| 310 mV]
[00:40:32.835,000] <inf> component_power: Current temperature: 32.212631 °C
[00:40:32.835,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2562|192] [4128 mV| 308 mV]
[00:40:42.936,000] <inf> component_power: Current temperature: 31.896545 °C
[00:40:42.937,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2562|194] [4128 mV| 312 mV]
[00:40:53.038,000] <inf> component_power: Current temperature: 32.212631 °C
[00:40:53.038,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2563|193] [4128 mV| 310 mV]
[00:41:03.139,000] <inf> component_power: Current temperature: 32.212631 °C
[00:41:03.140,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2561|196] [4126 mV| 314 mV]
[00:41:13.241,000] <inf> component_power: Current temperature: 31.896545 °C
[00:41:13.241,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2563|192] [4128 mV| 308 mV]
[00:41:23.343,000] <inf> component_power: Current temperature: 31.896545 °C
[00:41:23.343,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2563|193] [4128 mV| 310 mV]
[00:41:33.444,000] <inf> component_power: Current temperature: 31.896545 °C
[00:41:33.444,000] <dbg> component_power.power_work_handler: [BATT|MAIN] [2563|192] [4128 mV| 308 mV]
```

Reading is off though, room temperature should be ~ 28°C

Fixes #41742

### CODE REVIEW

**Comment by ycsin** [comment]
cc @wcappelle @Delsian

**Comment by ycsin** [comment]
@erwango I wonder what would happen if there are multiple threads/work that use the same ADC? All of them will setup the channel and then do `adc_read`. What happens if an interrupt occurs after the 1st thread setup the channel, and after servicing that interrupt, the scheduler schedules the other thread to run which setup & read the ADC in one go. Finally the 1st thread is scheduled and resume to continue from `adc_read`, but this time it would be reading the wrong channel configuration? @andyross Is what I described here possible?

Do you think we need a mutex lock in the `adc_channel_setup` & unlock at the end of `adc_read`, so that we can guarantee that the same thread will setup and read the ADC before another thread uses the ADC? The downside is that a thread will probably have to always setup the ADC before read, even if it is the only user of the entire application.

File: drivers/sensor/stm32_temp/stm32_temp.c [hunk 1]
```diff
@@ -160,6 +151,13 @@ static const struct stm32_temp_config stm32_temp_dev_config = {
 static struct stm32_temp_data stm32_temp_dev_data = {
 	.adc = DEVICE_DT_GET(DT_INST_IO_CHANNELS_CTLR(0)),
 	.channel = DT_INST_IO_CHANNELS_INPUT(0),
```
Comment (wcappelle) [inline drivers/sensor/stm32_temp/stm32_temp.c:153]: can be removed, seems duplicate to .channel_id in adc_cfg

**Comment by wcappelle** [comment]
> @erwango I wonder what would happen if there are multiple threads/work that use the same ADC? All of them will setup the channel and then do `adc_read`. What happens if an interrupt occurs after the 1st thread setup the channel, and after servicing that interrupt, the scheduler schedules the other thread to run which setup & read the ADC in one go. Finally the 1st thread is scheduled and resume to continue from `adc_read`, but this time it would be reading the wrong channel configuration? @andyross Is what I described here possible?
> 
> Do you think we need a mutex lock in the `adc_channel_setup` & unlock at the end of `adc_read`, so that we can guarantee that the same thread will setup and read the ADC before another thread uses the ADC? The downside is that a thread will probably have to always setup the ADC before read, even if it is the only user of the entire application.

Could you explain why you need to re-setup the channel every sample read? It's not clear to me in the description why this change is made. Since channel setup is done every sample read, it will also re-calibrate the adc every time. This looks like a bit overdoing it?

When I was testing the die temp sensor (on a STM32L5), the temperature was also a bit higher than the ambient temp, but I think this is because it's a bit heating during runtime. It's a die temperature sensor afterall. At cold startup, the temp of the die is about the ambient temp if I recall correctly. 
(Nevertheless, very jealous about your ambient temp :-) )

File: drivers/sensor/stm32_temp/stm32_temp.c (hunk 1 above)
Comment (ycsin) [reply drivers/sensor/stm32_temp/stm32_temp.c:153]: This `channel` variable in the `struct stm32_temp_data` indeed is not very useful, but this line is added by you. I can probably remove them altogether

**Comment by ycsin** [comment]
> Could you explain why you need to re-setup the channel every sample read? It's not clear to me in the description why this change is made. Since channel setup is done every sample read, it will also re-calibrate the adc every time. This looks like a bit overdoing it?

What if you have another thread that is using the ADC to sample another channel, when that thread configure the ADC, wouldn't it overwrite the configuration done by `stm32_temp`?



> Nevertheless, very jealous about your ambient temp :-)

It's an air-conditioned room, otherwise it would be 38 degree ☹️

File: drivers/sensor/stm32_temp/stm32_temp.c [hunk 2]
```diff
@@ -122,20 +121,11 @@ static int stm32_temp_init(const struct device *dev)
...
-					  .differential = 0 };
-	rc = adc_channel_setup(data->adc, accp);
-	LOG_DBG("Setup AIN%u got %d", data->channel, rc);
-
 	*asp = (struct adc_sequence){
-		.channels = BIT(data->channel),
+		.channels = BIT(data->adc_cfg.channel_id),
 		.buffer = &data->sample_buffer,
 		.buffer_size = sizeof(data->sample_buffer),
 		.resolution = 12,
```
Comment (ycsin) [inline drivers/sensor/stm32_temp/stm32_temp.c:128]: 12U

File: drivers/sensor/stm32_temp/stm32_temp.c (hunk 1 above)
Comment (ycsin) [reply drivers/sensor/stm32_temp/stm32_temp.c]: I have removed the `channel` variable from the struct

**Comment by ycsin** [comment]
@wcappelle could you have a look again?

**Review by wcappelle — DISMISSED** [review]

**Comment by ycsin** [comment]
@FRASTM could you review this patch? I've tested this on my 2.7 branch

File: dts/arm/st/g0/stm32g0b1.dtsi [hunk 3]
```diff
@@ -120,6 +120,19 @@
 		};
 	};
 
+	die_temp: dietemp {
```
Comment (FRASTM) [inline dts/arm/st/g0/stm32g0b1.dtsi:123]: this could be a node in the stm32g0 DTS

**Review by FRASTM — DISMISSED** [review]

File: dts/arm/st/g0/stm32g0b1.dtsi (hunk 3 above)
Comment (ycsin) [reply dts/arm/st/g0/stm32g0b1.dtsi:123]: > STM32G0B1 datasheet: image
> 
> I think the configurations should be similar across G0 family, but the TS_CAL2 seems to be missing from other series (at least G0B0 & G070), so I decided to leave them out from this PR for now.

Hmm, you sure?

Comment (FRASTM) [reply dts/arm/st/g0/stm32g0b1.dtsi:123]: ok, that's all surprising as the refMan of the stm32g0x0 explicitly mentions TS_CAL1 and TS_CAL2 where _TS_CAL2 is the temperature sensor calibration value acquired at TS_CAL2_TEMP
(refer to the datasheet for TS_CAL2 value)_  that does not exist

Comment (FRASTM) [reply dts/arm/st/g0/stm32g0b1.dtsi:123]: according to https://community.st.com/s/question/0D53W00000LFDS0SAP/stm32g070-how-to-calculate-the-current-ambient-temperature-from-internal-temperature-sensor
the formula seems to be : 
TS_adc_g0x0

File: drivers/sensor/stm32_temp/stm32_temp.c [hunk 4]
```diff
@@ -60,6 +59,9 @@ static int stm32_temp_sample_fetch(const struct device *dev, enum sensor_channel
 
 	k_mutex_lock(&data->mutex, K_FOREVER);
 
+	rc = adc_channel_setup(data->adc, &data->adc_cfg);
```
Comment (ycsin) [inline drivers/sensor/stm32_temp/stm32_temp.c:62]: Looks like I should return on error here instead of continuing

**Review by FRASTM — DISMISSED** [review]

File: drivers/sensor/stm32_temp/stm32_temp.c [hunk 5]
```diff
@@ -60,15 +59,21 @@ static int stm32_temp_sample_fetch(const struct device *dev, enum sensor_channel
...
+	if (rc) {
+		LOG_DBG("Setup AIN %u got %d", data->adc_cfg.channel_id, rc);
```
Comment (ycsin) [inline drivers/sensor/stm32_temp/stm32_temp.c:64]: If I edit this like so I will be able to save a few bytes if the user also compiled the vbat driver:

```suggestion
		LOG_DBG("Setup AIN%u got %d", data->adc_cfg.channel_id, rc);
```

**Comment by ycsin** [comment]
@erwango could you review this please?
@FRASTM could you review again please?

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Pull request merged**